### PR TITLE
Refactor S3 code to use the django storage API

### PR DESCRIPTION
**Pull request by omarkhan**
**Background:** xqueue currently uses S3 directly via the boto library. For deployments outside of AWS it would make sense to use a different cloud storage provider, or even the local filesystem.

In particular, OpenStack deployments will want to use an OpenStack Swift Object Store for user-uploaded content. Although Swift offers S3 API compatibility as a plugin (called "Swift3") in theory, in practice the compatibility is limited, often not installed on the OpenStack instance, and hard to get working.

This pull request is an attempt to solve this problem by refactoring all direct uses of the boto library to use the [django file storage API](https://docs.djangoproject.com/en/1.8/ref/files/storage/). The S3 backend from [django-storages](https://github.com/jschneier/django-storages) is included for backward-compatibility.

**Testing:** File uploads are completely untested as far as I can tell. I have added some tests for the submit handler.

See the description of https://github.com/edx/edx-platform/pull/11286 for instructions on how to test manually.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @omarkhan! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fxqueue&number=103

**Comment by Kelketek** [comment]
@omarkhan Is the way to test this by running it in tandem with the other PR? Since the test coverage is thin, as you say, can you make a test or two of your own?

**Comment by omarkhan** [comment]
@Kelketek I have added some tests in 40b8c9a

**Comment by Kelketek** [comment]
@omarkhan Needs a rebase. Is this installed on the sandbox? How do I verify it? By doing the things mentioned in the main PR? Does that invoke this code as part of it?

**Comment by Kelketek** [comment]
@omarkhan This needs a rebase, but it does work for me. :+1: if the rebase requires no major changes.

**Comment by omarkhan** [comment]
@Kelketek rebased

**Comment by edx-webhook** [comment]
Thanks for the pull request, @omarkhan! I've created [OSPR-1123](https://openedx.atlassian.net/browse/OSPR-1123) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/xqueue/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

**Comment by nedbat** [comment]
I've deleted the OSPR issue.

**Comment by Kelketek** [comment]
@nedbat These contributions concerning OpenStack aren't for solutions or edx directly, but are for a third party client. The OSPRs were deliberately made by Omar.

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @omarkhan! I've created [OSPR-1127](https://openedx.atlassian.net/browse/OSPR-1127) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

If you like, you can add yourself to the [AUTHORS](https://github.com/edx/xqueue/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

**Comment by ormsbee** [comment]
@jbarciauskas, @nedbat: 👍  from me, but someone in devops needs to review this since they own and deploy xqueue.

**Comment by jbarciauskas** [comment]
@edx/devops fyi

**Comment by jbarciauskas** [comment]
Bump @edx/devops again - goes hand in hand with the other Swift PR

File: queue/models.py [hunk 1]
```diff
@@ -23,8 +23,8 @@ class Submission(models.Model):
     xqueue_body      = models.TextField()
 
     # Uploaded files
-    s3_keys = models.CharField(max_length=CHARFIELD_LEN_LARGE) # S3 keys for internal Xqueue use
-    s3_urls = models.CharField(max_length=CHARFIELD_LEN_LARGE) # S3 urls for external access
+    keys = models.CharField(max_length=CHARFIELD_LEN_LARGE, db_column='s3_keys') # keys for internal Xqueue use
```
Comment (feanil) [inline queue/models.py]: I assume this doesn't generate any migrations since there is no actual change to the DB schema, but wanted to confirm.

**Comment by feanil** [comment]
Code looks good to me. 👍  just like with the edx-platform PR we should wait till the shared config PR is good to go(it's pretty close) before we merge this and that together.

File: queue/models.py (hunk 1 above)
Comment (omarkhan) [inline queue/models.py]: I just checked, turns out it does autogenerate a migration, even though there are no schema changes. Worse, when I run the migration it actually creates a new table, populates it, drops the old one and renames the new one. That could be very expensive if there is a lot of data, so I have reverted back to the original field names and added aliases instead: e08a1e4

**Comment by omarkhan** [comment]
Rebased

**Comment by jibsheet** [comment]
@bradenmacdonald Apparently this breaks submissions of grades for some external graders.
After this went live on Monday this week, it broke several teams (but not others).  I'll file a ticket once we have more details, but this will need to be triaged so that this code can be deployed.

**Comment by bradenmacdonald** [comment]
Thanks for the heads up @jibsheet. @e-kolpakov is on call this week for anything urgent - please tag him directly (as well as me) for faster response.

**Comment by jbarciauskas** [comment]
@bradenmacdonald @jibsheet @e-kolpakov what's the status on this?

**Comment by jibsheet** [comment]
@jbarciauskas Blocked on the course team - I did my part (setting up stage for testing) last Thursday.
Follow https://openedx.atlassian.net/projects/DEVOPS/queues/custom/16/DEVOPS-4691

**Comment by jbarciauskas** [comment]
Got it, thanks

**Pull request merged**